### PR TITLE
Plugins:  improve error message when plugin start fails

### LINKED ISSUE #8271: Plugins:  improve diagnosability of plugin launch failure
If a plugin fails on launch --- say, because of an unhandled exception --- or if a plugin times out during handshake, the error message ("A task was canceled.") [here](https://github.com/NuGet/NuGet.Client/blob/7fee43eeb1b5dbe0de3cb93ad0cb80b3a321e4a9/src/NuGet.Core/NuGet.Protocol/Plugins/PluginFactory.cs#L242) is unhelpful.

### PR DESCRIPTION
**Pull request by dtivel**
Fix https://github.com/NuGet/Home/issues/8271.

Improve the error message from "A task was canceled." to "Plugin 'x' failed within y seconds with exit code z."

### CODE REVIEW

File: src/NuGet.Core/NuGet.Protocol/Plugins/PluginFactory.cs [hunk 1]
```diff
@@ -235,6 +239,17 @@ public void Dispose()
...
+            catch (TaskCanceledException ex)
+            {
+                stopwatch.Stop();
+
+                int? exitCode = GetExitCodeOrNull(process);
+
+                plugin.Dispose();
+
+                throw new PluginException(
+                    string.Format(CultureInfo.CurrentCulture, Strings.Plugin_FailedOnCreation, plugin.Name, stopwatch.Elapsed.TotalSeconds, exitCode), ex);
```
Comment (jeffkl) [inline src/NuGet.Core/NuGet.Protocol/Plugins/PluginFactory.cs]: `TotalSeconds` is a `double`, you might want to round it off in the string resource 

`Plugin '{0}' failed within {1:N2} seconds with exit code {2}.`

File: src/NuGet.Core/NuGet.Protocol/Plugins/PluginFactory.cs [hunk 2]
```diff
@@ -426,6 +441,20 @@ private void UnregisterEventHandlers(Plugin plugin)
             }
         }
 
+        private static int? GetExitCodeOrNull(Process process)
```
Comment (jeffkl) [inline src/NuGet.Core/NuGet.Protocol/Plugins/PluginFactory.cs]: Extension method instead?

**Review by jeffkl — APPROVED** [review]

File: src/NuGet.Core/NuGet.Protocol/Plugins/PluginFactory.cs (hunk 2 above)
Comment (dtivel) [reply src/NuGet.Core/NuGet.Protocol/Plugins/PluginFactory.cs]: If there were a general need for this in the future, maybe.  Not now.

**Review by zivkan — APPROVED** [review]

File: src/NuGet.Core/NuGet.Protocol/Plugins/Plugin.cs [hunk 3]
```diff
@@ -157,9 +157,10 @@ public void Dispose()
                 _process.Exited -= OnExited;
 
                 _process.Kill();
-                _process.Dispose();
             }
 
+            _process.Dispose();
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Protocol/Plugins/Plugin.cs:162]: Dispose the `System.Diagnostics.Process` object whether it represents the current process or a plugin process.  Previously, it wasn't being disposed when it represented the current process.

File: src/NuGet.Core/NuGet.Protocol/Plugins/PluginFactory.cs [hunk 4]
```diff
@@ -216,25 +213,44 @@ public void Dispose()
...
 
-                process.EnableRaisingEvents = true;
+                // It's critical that this be registered after ConnectAsync(...).
+                // If it's registered before, stuff could be disposed, which would lead to unexpected exceptions.
+                // If the plugin process exited before this event is registered, an exception should be caught below.
+                plugin.Exited += OnPluginExited;
             }
             catch (ProtocolException ex)
             {
+                plugin.Dispose();
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Protocol/Plugins/PluginFactory.cs:234]: A plugin instance wasn't previously disposed in this rare, exceptional case.

**Review by nkolev92 — APPROVED** [review]
LGTM. 

I think the risk is low, so I think it's ok to try to take in 5.2.

**Pull request merged**